service/controller/resource/clusterconditions/conditionupgrading.go

```diff
@@ -2,165 +2,95 @@ package clusterconditions
 
 import (
 	"context"
-	"time"
 
 	"github.com/blang/semver"
+	"github.com/giantswarm/apiextensions/v3/pkg/annotation"
 	aeconditions "github.com/giantswarm/apiextensions/v3/pkg/conditions"
 	"github.com/giantswarm/microerror"
-	corev1 "k8s.io/api/core/v1"
 	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
 	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
 
 	"github.com/giantswarm/azure-operator/v5/pkg/conditions"
+	"github.com/giantswarm/azure-operator/v5/pkg/upgrade"
 	"github.com/giantswarm/azure-operator/v5/service/controller/key"
 )
 
 func (r *Resource) ensureUpgradingCondition(ctx context.Context, cluster *capi.Cluster) error {
+	if conditions.IsUnexpected(cluster, aeconditions.UpgradingCondition) {
+		return microerror.Maskf(
+			conditions.UnexpectedConditionStatusError,
+			conditions.UnexpectedConditionStatusErrorMessage(cluster, aeconditions.UpgradingCondition))
+	}
+
 	// Case 1: new cluster just being created, no upgrade yet.
 	if capiconditions.IsTrue(cluster, aeconditions.CreatingCondition) {
-		err := conditions.MarkUpgradingNotStarted(cluster)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-
+		conditions.MarkUpgradingNotStarted(cluster)
 		return nil
 	}
 
-	// Case 2: Upgrading=False, cluster is currently not in upgrading state,
-	// let's check if it should be.
-	if capiconditions.IsFalse(cluster, aeconditions.UpgradingCondition) {
-		err := r.checkIfUpgradingHasBeenStarted(ctx, cluster)
-		if err != nil {
-			return microerror.Mask(err)
+	// Case 2: first upgrade to node pools release, a bit of an edgecase,
+	// albeit an important one :)
+	if upgrade.IsFirstNodePoolUpgradeInProgress(cluster) {
+		if !capiconditions.IsTrue(cluster, aeconditions.UpgradingCondition) {
+			conditions.MarkUpgradingStarted(cluster)
```

Comment:
Here we are checking if the cluster is being upgraded to the node pools release for the first time, the check is using new `release.giantswarm.io/upgrading-to-node-pools` annotation. (we can remove this as soon as all cluster have been upgraded)